File: .cpp/51.N-Queens.cpp
class Solution
{
public:
    void solve(int col, vector<string> &board, vector<vector<string>> &result,
               vector<int> &rowArray, vector<int> &mainDiagonal,
               vector<int> &crossDiagonal, int n)
    {
        if (col == n)
        {
            result.push_back(board);
            return;
        }

        for (int row = 0; row < n; row++)
        {
            if (rowArray[row] == 0 && mainDiagonal[row + col] == 0 && crossDiagonal[n - 1 + col - row] == 0)
            {
                rowArray[row] = 1;
                mainDiagonal[row + col] = 1;
                crossDiagonal[n - 1 + col - row] = 1;
                board[row][col] = 'Q';
                solve(col + 1, board, result, rowArray, mainDiagonal, crossDiagonal, n);
                rowArray[row] = 0;
                mainDiagonal[row + col] = 0;
                crossDiagonal[n - 1 + col - row] = 0;
                board[row][col] = '.';
            }
        }
    }
    vector<vector<string>> solveNQueens(int n)
    {
        vector<vector<string>> result;
        vector<string> board(n);
        string s(n, '.');
        for (int i = 0; i < n; i++)
        {
            board[i] = s;
        }
        vector<int> rowArray(n, 0), mainDiagonal(2 * n - 1, 0),
            crossDiagonal(2 * n - 1, 0);
        solve(0, board, result, rowArray, mainDiagonal, crossDiagonal, n);
        return result;
    }
};

File: 51.N-Queens.py
class Solution:
    def solve(self, col, board, result, rowArray, majorDiagonal, minorDiagonal, n):
        if col == n:
            result.append(board[:])
            return

        for row in range(n):
            if (
                rowArray[row] == 0
                and majorDiagonal[row + col] == 0
                and minorDiagonal[n - 1 + col - row] == 0
            ):
                rowArray[row] = 1
                majorDiagonal[row + col] = 1
                minorDiagonal[n - 1 + col - row] = 1
                board[row] = board[row][:col] + "Q" + board[row][col + 1 :]
                self.solve(
                    col + 1, board, result, rowArray, majorDiagonal, minorDiagonal, n
                )
                rowArray[row] = 0
                majorDiagonal[row + col] = 0
                minorDiagonal[n - 1 + col - row] = 0
                board[row] = board[row][:col] + "." + board[row][col + 1 :]

    def solveNQueens(self, n: int) -> List[List[str]]:
        result = []
        board = ["." * n for _ in range(n)]
        rowArray = [0] * n
        majorDiagonal = [0] * (2 * n - 1)
        minorDiagonal = [0] * (2 * n - 1)
        self.solve(0, board, result, rowArray, majorDiagonal, minorDiagonal, n)
        return result
